### ISSUE TITLE
Custom match for operations with arguments

### ISSUE THREAD

**Issue by Necr0x0Der** [trueagi-io/hyperon-experimental#546]
Currently, grounded atoms can have custom `match_`, but it works for atoms themselves, while it would be useful for operations to handle their arguments in `match_` as well. For example, `(+ $x 1)` could handle matching against `5` inferring a binding `$x <- 4`. Similarly, `cons-atom` could be used for deconstruction, e.g. `(cons-atom $car $cdr)` matched against `(a b c)` would yield bindings `$car <- a`, `$cdr <- (b c)`, so one could write
```
(case $x
  ((cons-atom $car $cdr) ...)
```
instead of
```
(let ($car $cdr) ((car-atom $x) (cdr-atom $x)) ...
```
in recursive analysis of expressions.

It is generally not useful to have a custom match for operations without arguments.

**Comment by Necr0x0Der** [comment]
This might even be usable in function definitions, e.g.
```
(= (is-closed (cons-atom $x $xs)) (and (is-closed $x) (is-closed $xs)))
```

**Comment by ngeiswei** [comment]
Generic functional pattern matching is costly of course.  I wonder if we could simply, in the meantime, merely support it for `cons-atom`, a bit like in traditional functional programming languages, where `(A B C)` could be viewed as syntactic sugar for `(cons-atom A (cons-atom B C))`, regardless of its internal representation.

**Comment by Necr0x0Der** [comment]
> we could allow writing custom functional pattern implementations for the grounded atoms

Yes. In this case, grounded operations could be used for deconstruction in the same way as it is done for symbolic constructors. I guess this will be necessary for the compiler, in which there is less distinction between grounded and symbolic atoms. But we can still consider doing this in the interpreter as well
